### DISCUSSION TITLE
Should we incorporate {renv} into this template?

### DISCUSSION THREAD

**Discussion by bpffjl** [GSA/msg_project_template#1] [Ideas]
I was think about virtual environments and stumbled across [renv](https://rstudio.github.io/renv/articles/renv.html). I've never used it independently, but @jwbowers executed a script on the power analysis that we were working on together. What do you (@saraguenther and @jwbowers) think?

**Comment by jwbowers** [comment]
I really like renv. It can be fiddly, but it is better than anything else I've tried to ensure that (1) all collaborators are using the same packages and (2) the same versions.

I would advocate that we use it.

renv and here are two super pieces of workflow.

The last piece is a Makefile, I'm intrigued by snakemake <https://lachlandeer.github.io/snakemake-econ-r-tutorial/> but haven't tried it out yet.

**Reply by bpffjl** [reply]
Thanks Jake. Sometimes my head gets spinning on these types of processes. I went down a small rabbit hole on [Makefiles](https://www.gnu.org/software/make/manual/html_node/Introduction.html) and landed on the [{targets}](https://docs.ropensci.org/targets/index.html) packages. I think that I understand the utility of this process, but I'm not clear how to implement it on our projects. 

For renv, we should add something about this to the SOP that we're going to develop. Also, would it be useful to write some template scripts to set up renv and add code to other templates? In addition to the SOP should we add something to the readme file?

**Reply by jwbowers** [reply]
Now that I've read about the targets package, I'll be very interested to see what you make of it @bpffjl . My first look suggests that I'd have to change how I code within R to use it. But, maybe I'm wrong about it. And perhaps changing how I code (so that everything is a function) might be best for our work. With Makefiles, FWIW, they are implemented in the root directory and just list the inputs and outputs.

**Reply by bpffjl** [reply]
I agree. The targets structure appears to be all functions all the time. I don't love that, because I'm a fan of the markdown structure for all steps because they produce records of the process you've undertaken and are great for narrative and notes. So I'm

In a way, I make something akin to Makefiles, because I often write R files to render my Rmarkdown files. It's not as complete or clear as a Makefile so I think that it doesn't meet the need completely. 

Have you used a package in the past to create your Makefiles? I've seen one called Makefiler.

**Reply by jwbowers** [reply]
I just edit the Makefile directly and don't do so many fancy things in it. Mostly I try to edit it whenever I make a new file and/or save a new output. Basically, I work on so many projects that I'm constantly afraid I'll forget how things go together.

Once in a while I use a macro like

```make
RCMD = R -f

## Any DA possible but only among non-white respondents
dist_mats_anyDA_VM.rda: dist_mats_data_anyDA_VM.R wrkdatOwnMap.rda nonbimatchingfunctions.R
        $(RCMD) dist_mats_data_anyDA_VM.R
```

The problem I see right now is that it is hard to install make on Windows. It looks like Chocolatey is approved by the GSA and I'm hoping it is installed on my machine today where I can type [`choco install make`](https://community.chocolatey.org/packages/make)  If that doesn't work, I might suggest [snakemake](https://snakemake.readthedocs.io/en/stable/tutorial/basics.html) because again, it can run anything (so, a project could have a python script, an R script, an R markdown file, a bash command, etc..).  

And I agree that targets seems too heavy a lift to make every action into a function.

We could also just do it within R and say in the README: "First run the Make.R file by typing Rscript ....". Or something.

**Reply by jwbowers** [reply]
I also just found [nextflow](https://nextflow.io/example5.html). Still a bit unclear about how it lives and interacts.

**Comment by jwbowers** [comment]
The targets package looks cool. Eager to try it.

renv doesn't really require files. After installing the library renv,  It just requires you to do:  (1) `renv::init()` and then following the instructions (basically, typing `renv::snapshot()` to save the list of packages after you install or update any new packages.  and (2) add renv.lock and renv/activate and .Rprofile to github. (At least I think so).

**Reply by bpffjl** [reply]
When do you run `renv::init()`? When first setting up a project? Do you run it again if you need a new package?

**Comment by jwbowers** [comment]
You can see a repo using renv here: <https://github.com/GSA/oes-msg-voting_description>

Once you have some code (like, say, `Scripts/Analysis/data_cleaning.R`,
that has lines like

```r
library(tidyverse)
library(estimatr)
```

Then if you start R in the root directory and renv is not yet installed and
active, you need type `renv::init()` and renv will search through the files
for `library(...` commands and record those packages.

You can see more here: https://rstudio.github.io/renv/articles/renv.html


Jake Bowers
http://jakebowers.org
